api/src/wfl/executor.clj

```diff
@@ -428,6 +446,80 @@
+               WHERE submission IN
+                 (SELECT DISTINCT submission FROM %s
+                 WHERE workflow IN %s)"]
+    (->> (format query details details workflow-ids)
+         (jdbc/query tx)
+         (group-by :reference))))
+
+(defn update-retried-workflows
+  "Match `original-records` with `retry-records` on entity uuid,
+  update each original record to point to its retry record,
+  and propagate all updates to `details` table."
+  [{:keys [details] :as executor} [retry-records original-records]]
+  (letfn [(link-retry-to-original
+            [{:keys [entity]         :as original}
+             {:keys [retryEntity id] :as retry}]
+            (when-not (= entity retryEntity)
+              (throw (ex-info (str "Failed to link retry to original workflow: "
+                                   "mismatched entities")
+                              {:original original
+                               :retry    retry
+                               :executor executor})))
+            (assoc original :retry id))
+          (write-retries
+            [now records]
+            (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+              (run!
+               (fn [{:keys [id retry]}]
+                 (jdbc/update! tx details
+                               {:retry retry :updated now}
+                               ["id = ?" id]))
+               records)))]
+    (when-not (== (count retry-records) (count original-records))
+      (throw (ex-info "All original records should be retried"
+                      {:retry-records    retry-records
+                       :original-records original-records
+                       :executor         executor})))
+    (log/debug (format "%s Linking %s retry records to their originals..."
+                       (log-prefix executor) (count retry-records)))
+    (->> retry-records
+         (map #(rename-keys % {:entity :retryEntity}))
+         (sort-by :retryEntity)
+         (map link-retry-to-original (sort-by :entity original-records))
+         (write-retries (utc-now)))))
+
+(defn ^:private retry-terra-executor
+  "Resubmit the entities associated with `workflows` in `workspace`
+  and update each original workflow with the row ID of its retry."
+  [{:keys [workspace] :as executor} workflows]
+  ;; TODO: need some sort of fromSource check/coercion:
+  ;; may not always be dealing with references, should handle generic entity.
+  (letfn [(retried-workflow-records [reference-id]
+            (let [reference (rawls/get-snapshot-reference workspace reference-id)]
+              (->> reference
+                   (create-submission! executor)
+                   (allocate-submission executor reference))))]
+    (->> (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+           (reference-id-to-workflow-records tx executor workflows))
+         (map-keys retried-workflow-records)
+         (run! (partial update-retried-workflows executor)))))
```

Comment:
I got burned by this too!  Initially used a `map` which yields a lazy sequence.  When I cleaned up debug printing, this was no longer called.  We're running the function for its side effects rather than its returned value.